File: src/main/java/exnihiloomnia/client/particles/ParticleSieve.java
package exnihiloomnia.client.particles;

import net.minecraft.client.particle.Particle;
import net.minecraft.client.renderer.texture.TextureAtlasSprite;
import net.minecraft.world.World;

public class ParticleSieve extends Particle{
	protected float minU = 0.0f;
	protected float minV = 0.0f;
	protected float maxU = 0.0f;
	protected float maxV = 0.0f;
	
	public ParticleSieve(World par1World, double par2, double par4, double par6, double par8, double par10, double par12, TextureAtlasSprite texture)
    {
        super(par1World, par2, par4, par6, par8, par10, par12);
        this.particleTexture = texture;
        this.particleRed = this.particleGreen = this.particleBlue = 1.0F;
        this.setSize(0.04F, 0.04F);
        this.particleScale = 0.07f + (this.rand.nextFloat() * 0.1F);
        this.motionX *= 0.019999999552965164D;
        this.motionY *= -0.4;
        this.motionZ *= 0.019999999552965164D;
        this.particleMaxAge = (int)(20.0D / (Math.random() * 0.2D + 0.2D));
        
        
        //Select a single pixel in the texture to draw.
        float width = (this.particleTexture.getMaxU() - this.particleTexture.getMinU()) / particleTexture.getIconWidth();
        float height = (this.particleTexture.getMaxV() - this.particleTexture.getMinV()) / particleTexture.getIconHeight();
        
        this.minU = this.particleTexture.getMinU() + (rand.nextInt(particleTexture.getIconWidth()) * width);
        this.minV = this.particleTexture.getMinV() + (rand.nextInt(particleTexture.getIconHeight()) * height);
        this.maxU = this.minU + width;
        this.maxV = this.minV + height;
    }

	@Override
    public void onUpdate()
    {
        this.prevPosX = this.posX;
        this.prevPosY = this.posY;
        this.prevPosZ = this.posZ;
        this.moveEntity(this.motionX, this.motionY, this.motionZ);
        this.motionX *= 0.99D;
        this.motionY *= 0.99D;
        this.motionZ *= 0.99D;
        
        this.nextTextureIndexX();

        if (this.particleMaxAge-- <= 0)
        {
            this.setExpired();
        }
    }
    
    @Override
    public int getFXLayer()
    {
        return 1;
    }
}

File: src/main/kotlin/com/jozufozu/exnihiloomnia/client/ParticleSieve.kt
package com.jozufozu.exnihiloomnia.client

import net.minecraft.block.Block
import net.minecraft.client.particle.ParticleDigging
import net.minecraft.item.ItemStack
import net.minecraft.util.math.BlockPos
import net.minecraft.world.World
import kotlin.math.ceil

class ParticleSieve(world: World, mimic: ItemStack, sievePos: BlockPos)
    : ParticleDigging(
        world,
        sievePos.x.toDouble() + world.rand.nextDouble() * 0.875 + 0.0625,
        sievePos.y + 8.5 / 16.0,
        sievePos.z.toDouble() + world.rand.nextDouble() * 0.875 + 0.0625,
        0.0, 0.0, 0.0,
        Block.getBlockFromItem(mimic.item).getStateFromMeta(mimic.metadata)) {
    init {
        setBlockPos(sievePos)

        motionX = 0.0
        motionY = 0.0
        motionZ = 0.0

        particleScale /= 5.0f
        particleMaxAge = ceil(particleMaxAge.toDouble() * 1.3).toInt()
    }

    override fun onUpdate() {
        prevPosX = posX
        prevPosY = posY
        prevPosZ = posZ

        if (particleAge++ >= particleMaxAge) {
            setExpired()
        }

        motionY -= 0.002 * particleGravity.toDouble()
        move(motionX, motionY, motionZ)

        motionY *= 0.9800000190734863

        if (onGround) {
            motionX *= 0.699999988079071
            motionZ *= 0.699999988079071
        }
    }
}
